### PR TITLE
Use 64bits integers for offsets inside scorpio interface

### LINKED ISSUE #1846: segfault with ne1024 and pg2 (on pm-cpu with gnu)
With ne1024pg2, I have hit a few segfaults. The first thing I tried was using more nodes, but I'm leaning against the possibility of there not being enough memory. I can run ne120pg2 with 6 nodes of pm-cpu and ne256pg2 with 24 nodes.  I've hit segfault running ne1024 case with 256,384,512,and 768 nodes. I also tried DEBUG. If we can believe the stack traces, I think there might be at least two different places it has stopped. 

The latest IC file: `/global/cfs/cdirs/e3sm/inputdata/atm/scream/init/screami_ne1024np4L128_20220805.nc`

One stopped location using 384 nodes and DEBUG:
```
4615: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
4615: 
4615: Backtrace for this error:
4615: #0  0x15214f7dfd6f in ???
4615: #1  0x29537be in __scream_scorpio_interface_MOD_eam_pio_closefile
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scream_scorpio_interface.F90:752
4615: #2  0x2952f1c in __scream_scorpio_interface_MOD_eam_pio_finalize
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scream_scorpio_interface.F90:889
4615: #3  0x2952ec7 in errorhandle
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scream_scorpio_interface.F90:930
4615: #4  0x2957af5 in __scream_scorpio_interface_MOD_get_variable
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scream_scorpio_interface.F90:399
4615: #5  0x295aa87 in get_variable_c2f
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scream_scorpio_interface_iso_c2f.F90:139
4615: #6  0x2959e08 in _ZN6scream7scorpio12get_variableERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES8_S8_iRKSt6vectorIS6_SaIS6_EEiS8_
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scream_scorpio_interface.cpp:85
4615: #7  0x297553b in _ZN6scream15AtmosphereInput18register_variablesEv
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scorpio_input.cpp:432
4615: #8  0x2975403 in _ZN6scream15AtmosphereInput23init_scorpio_structuresEv
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scorpio_input.cpp:403
4615: #9  0x29713b7 in _ZN6scream15AtmosphereInput4initERKSt10shared_ptrIKNS_12FieldManagerEERKS1_IKNS_12GridsManagerEE
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scorpio_input.cpp:62
4615: #10  0x2971271 in _ZN6scream15AtmosphereInputC2ERKN4ekat13ParameterListERKSt10shared_ptrIKNS_12FieldManagerEERKS5_IKNS_12GridsManagerEE
4615:   at /global/cfs/cdirs/e3sm/ndk/se14-aug5/components/scream/src/share/io/scorpio_input.cpp:37

/pscratch/sd/n/ndk/e3sm_scratch/pm-cpu/se14-aug5/f1024.F2010-SCREAMv1-noAero.ne1024pg2_r0125_oRRS18to6v3.se14-aug5.gnu.24s.n0384a16x8c8il4.ICaug5.DEBUG
```

`components/scream/src/share/io/scream_scorpio_interface.F90`
```
    ! Find the pointer for this file                                                                                                                                                                                                                     
    call lookup_pio_atm_file(trim(fname),pio_atm_file,found,pio_file_list_ptr)
    if (found) then
      if (pio_atm_file%num_customers .eq. 1) then
        if (pio_atm_file%purpose .eq. file_purpose_out) then
          call PIO_syncfile(pio_atm_file%pioFileDesc)
        endif
        call PIO_closefile(pio_atm_file%pioFileDesc)
        pio_atm_file%num_customers = pio_atm_file%num_customers - 1

        ! Remove all variables from this file as customers for the stored pio                                                                                                                                                                            
        ! decompostions                                                                                                                                                                                                                                  
        curr_var_list => pio_atm_file%var_list_top  ! Start with the first variable in the file                                                                                                                                                          
        do while (associated(curr_var_list))
          var => curr_var_list%var  ! The actual variable pointer                                                                                                                                                                                        
          if (associated(var)) then
            ! Remove this variable as a customer of the associated iodesc                                                                                                                                                                                
            var%iodesc_list%num_customers = var%iodesc_list%num_customers - 1   ! <-- line 752
            ! Dellocate select memory from this variable.  Note we can't just                                                                                                                                                                            
            ! deallocate the whole var structure because this would also                                                                                                                                                                                 
            ! deallocate the iodesc_list.                                                                                                                                                                                                                
            call deallocate_hist_var_t(var)
          end if ! associated(var)                                                                                                                                                                                                                       
          curr_var_list => curr_var_list%next  ! Move on to the next variable                                                                                                                                                                            
        end do ! associated(curr_var_list)                                                                                                                                                                                                               

```

The other location I see cases stop on is here (in this case 768 nodes without DEBUG):

```
 3837: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
 3837: 
 3837: Backtrace for this error:
 3837: #0  0x152c98243d6f in ???
 3837: #1  0x102b7aa in ???
 3837: #2  0x102b9dc in ???
 3837: #3  0x1030699 in ???
 3837: #4  0x1032a96 in ???
 3837: #5  0x10315dd in ???
 3837: #6  0x1043396 in ???
 3837: #7  0x10497b7 in ???
 3837: #8  0x1049c33 in ???
 3837: #9  0xf6f0b3 in ???
 3837: #10  0xf72889 in ???
 3837: #11  0xf74425 in ???
 3837: #12  0x553f17 in ???
 3837: #13  0x54f573 in ???
 3837: #14  0x4ccf64 in __component_mod_MOD_component_init_cc
 3837:  at /global/cfs/cdirs/e3sm/ndk/se14-aug5/driver-mct/main/component_mod.F90:258
 3837: #15  0x4bd2da in __cime_comp_mod_MOD_cime_init
 3837:  at /global/cfs/cdirs/e3sm/ndk/se14-aug5/driver-mct/main/cime_comp_mod.F90:1438
 3837: #16  0x492f0c in cime_driver
 3837:  at /global/cfs/cdirs/e3sm/ndk/se14-aug5/driver-mct/main/cime_driver.F90:122
 3837: #17  0x492f0c in main
 3837:  at /global/cfs/cdirs/e3sm/ndk/se14-aug5/driver-mct/main/cime_driver.F90:23

/pscratch/sd/n/ndk/e3sm_scratch/pm-cpu/se14-aug5/f1024.F2010-SCREAMv1-noAero.ne1024pg2_r0125_oRRS18to6v3.se14-aug5.gnu.24s.n0768a32x4c4il2.ICaug5
```

`driver-mct/main/component_mod.F90`
```
    do eci = 1,size(comp)
       if (iamroot_CPLID .and. comp(eci)%present) then
          write(logunit,F00) 'Initialize component '//trim(comp(eci)%ntype)
          call shr_sys_flush(logunit)
       endif

       if (.not. associated(comp(eci)%x2c_cc)) allocate(comp(eci)%x2c_cc)
       if (.not. associated(comp(eci)%c2x_cc)) then
          allocate(comp(eci)%c2x_cc)
          ! this is needed for check_fields                                                                                                                                              
          nullify(comp(eci)%c2x_cc%rattr)
       endif
       if (comp(eci)%iamin_compid .and. comp(eci)%present) then
          if (drv_threading) call seq_comm_setnthreads(comp(eci)%nthreads_compid)
          call shr_sys_flush(logunit)

          if (present(seq_flds_x2c_fluxes)) then
             call mct_avect_vecmult(comp(eci)%x2c_cc, comp(eci)%drv2mdl, seq_flds_x2c_fluxes, mask_spval=.true.)
          end if

          call t_startf('comp_init')
          call comp_init( EClock, comp(eci)%cdata_cc, comp(eci)%x2c_cc, comp(eci)%c2x_cc, &   ! <-- line 258
               NLFilename=NLFilename )
          call t_stopf('comp_init')
          if(nan_check_component_fields) then
```

### PR DESCRIPTION
**Pull request by bartgol**
With ne1024 we hit a problem with our scorpio interface, which uses integers (32 bits) for offsets in the global array. With ne1024, and 128 levels, we have more than 7B unknowns for 3d variables, which cause integer overflow.

Fixes #1846 (hopefully).

### CODE REVIEW

**Comment by ndkeen** [comment]
I can build it on pm-cpu, but I get error with ne30:
```
 19: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
 19: 
 19: Backtrace for this error:
 19: #0  0x1525c6a88d6f in ???
 19: #1  0x1525cea3c3bd in __freeBlasMemPool
 19:    at ./src/crayblas_util.c:353
 19: #2  0x1525cc8c5894 in __crayblas_shutdown
 19:    at ./src/crayblas_conf.c:232
 19: #3  0x1525d5aae5a2 in ???
 19: #4  0x1525c6a8bb08 in ???
 19: #5  0x1525c6a8bc99 in ???
 19: #6  0x1525c6a732c3 in ???
 19: #7  0x4149c9 in ???
 19:    at ../sysdeps/x86_64/start.S:120
```

**Comment by bartgol** [comment]
I'm building an ne30 case on mappy, to see if I can reproduce.

**Comment by bartgol** [comment]
On mappy, an SMS test with ne30np4 runs fine (10 timestep done so far). @ndkeen  I'm assuming the error is at init time, or at first timestep?

**Comment by ndkeen** [comment]
note that my error was from a simple rebuild in a case after these changes.  when i try a new case on cori, it seems to be working and i'm doing the same on pm-gpu.

File: components/scream/src/share/io/scream_scorpio_interface.hpp [hunk 1]
```diff
@@ -23,6 +23,8 @@ static constexpr int PIO_INT = 4;
 namespace scream {
 namespace scorpio {
 
+  using offset_t = std::int64_t;
```
Comment (jeff-cohere) [inline components/scream/src/share/io/scream_scorpio_interface.hpp:26]: Neither here nor there, but `size_t` and `ssize_t` (its signed equivalent) kind of serve the same purpose, if you had any reservations about adding a new type alias.

**Comment by ndkeen** [comment]
The issue this PR fixes had a few other problems, but I think it's fair to say this issue fixes it.

File: components/scream/src/share/io/scream_scorpio_interface.hpp (hunk 1 above)
Comment (bartgol) [reply components/scream/src/share/io/scream_scorpio_interface.hpp:26]: In F90 they use unsigned integers. If I want to pass a ptr to our offsets arrays to scorpio, I think I need unsigned as well.

**Review by jeff-cohere — APPROVED** [review]

**Pull request merged**